.github/workflows/build-and-test.yml

```
@@ -0,0 +1,63 @@
+name: Build and test
+on:
+  push:
+    paths:
+      - .github/workflows/**
+      - e2eTests/**
+      - tests/**
+      - todo_app/**
+      - Dockerfile
+      - poetry.lock
+  pull_request:
+    paths:
+      - .github/workflows/**
+      - e2eTests/**
+      - tests/**
+      - todo_app/**
+      - Dockerfile
+      - poetry.lock
+
+jobs:
+  build:
+    name: Build and test
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+
+    - name: Integration test build
+      run: docker build --target pipeline_integration_tests --tag todo-app:pipeline_integration_tests .
+
+    - name: Run integration tests
+      run: docker run todo-app:pipeline_integration_tests
+
+    - name: e2e tests build
+      run: docker build --target pipeline_e2e_tests --tag todo-app:pipeline_e2e_tests .
+
+    - name: Run e2e tests
+      run: >
+        docker run
+        -e FLASK_APP=todo_app/app
+        -e TRELLO_TO_DO_BOARD_ID=65cb37a378d7b6f10d62c208
+        -e TRELLO_ORGANIZATION_ID=65cb37900e446a315b3b1fbc
+        -e TRELLO_API_KEY=$TRELLO_API_KEY
+        -e TRELLO_API_TOKEN=$TRELLO_API_TOKEN
```

Very minor, but worth noting that Docker has a nice shorthand for this scenario where you want an environment variable that matches the same variable outside:
```suggestion
        -e TRELLO_API_KEY
        -e TRELLO_API_TOKEN
```

That said, I can't spot where these variables come from, which suggests to me that they might be being stored as variables rather than secrets - does that sound plausible?

I'd expect somewhere an env block like below (which can be set either at the job level or a specific step-level):
```
env:
    TRELLO_API_KEY: ${{ secrets.TRELLO_API_KEY }}
    TRELLO_API_TOKEN: ${{ secrets.TRELLO_API_TOKEN }}
```
I may have just overlooked something, but worth checking as if these are just variables then I think there's a risk that GitHub doesn't mask them if they were accidentally outputted during a run

---

Ah, so you're right that they're not pulling them from the secrets here, but I also haven't set them as plain environment variables, they're just not defined 🤣! Because I'm using vcr cassesstes for the e2e tests rather than actual calls to Trello it doesn't care if they're valid. But I can make that clearer by putting in dummy keys and tokens in the pipeline docker run command.